jobrunner/create_or_update_jobs.py

```diff
@@ -173,39 +165,12 @@ def get_latest_job_for_each_action(workspace):
     Return a list containing the most recent job (if any) for each action in
     the workspace
     """
-    # We treat the files on disk as the canonical source for the state of
-    # completed jobs.
-    completed_jobs = get_completed_jobs_from_disk(workspace)
-    # However active jobs aren't represented on disk so if we want to know what
-    # jobs are currently running we have to ask the database.
-    active_jobs = get_active_jobs_from_database(workspace)
-    # Combine the two sources of jobs. Where there is an active job and a
-    # completed job for the same action we prefer the active job as that is
-    # necessarily the more recent.
-    combined = {job.action: job for job in completed_jobs + active_jobs}
-    return list(combined.values())
-
-
-def get_active_jobs_from_database(workspace):
-    return find_where(
-        Job,
-        workspace=workspace,
-        state__in=[State.PENDING, State.RUNNING],
-    )
-
-
-def get_completed_jobs_from_disk(workspace):
-    # This is slightly inelegant but helps keep the rest of the code simple:
-    # the `manifest.json` file in the workspace directory on disk stores the
-    # final state of completed jobs, but doesn't store the full set of Job
-    # object fields because they aren't generally needed. This means we can't
-    # reconstruct Job instances to return so instead we create a JobPlaceholder
-    # containing just the two fields we care about in this context: the action
-    # and the final state.
-    return [
-        JobPlaceholder(action=action, state=state)
-        for action, state in get_states_for_actions(workspace).items()
-    ]
+    all_jobs = find_where(Job, workspace=workspace)
+    latest_jobs = []
+    for _, jobs in groupby(sorted(all_jobs, key=attrgetter("action")), key=attrgetter("action")):
```

Comment:
This might be a little easier to parse if you sorted on the line above the groupby?  The double key= on this line caught me out a bit at first despite being quite familiar with both both functions! (it's a shame there isn't a way to do both given you're using the same key)

Reply:
Is this more readable? I'm not sure it is!

```
key = attrgetter("action")
all_jobs.sort(key=key)
for _, jobs in groupby(all_jobs, key=key)
```

Reply:
I'd be tempted by this:
```py
latest_by_action = {job.action: job for job in sorted(all_jobs, key=attrgetter("created_at")}
return list(latest_by_action.values())
```
But I don't know if that's clearer to anyone except me :smile:

Reply:
I only meant to pull the `sorted(all_jobs, key=attrgetter("action"))` portion out as `key=attrgetter("action")), key=attrgetter("action"))` tripped me up a little bit 🙂

Reply:
My inclination (which I considered when originally writing it) is to pull out a helper function that implements the group by operation without requiring it to be pre-sorted.

Reply:
I don't like having the sort separate because it's purely an implementation detail of the group by, not something that we actually want to achieve.